active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,711 @@
+
+Here is an example:
+
+main.bicepparam
+
+```bicep
+using 'main.bicep'
+
+param boolParam = true
+// ...
+
+extensionConfigs {
+  k8s: {
+    ✅ kubeConfig: az.aksKubeConfig(...) // any data needed to get it. Overloads may depend on auth workflow.
+    ✅ kubeConfig: getSecret(...)
+    ❌ kubeConfig: loadFileContent('...')
+    ❌ kubeConfig: 'inlinedSecret'
+    namespace: 'default'
+  }
+}
+```
+
+If extensions require a configuration, there must be a parameters file (Bicep or JSON) that provide the extension
+configuration. If it is not supplied, there should be appropriate error diagnostics. There will be no support for
+inlining extension configurations from CLI arguments.
+
+### The new design from the ARM template perspective
+
+The main Bicep deployment would compile to the ARM template following this paragraph. Only REP relevant data is shown.
+The "extensionConfigs" deployment property will be similar in style to how deployment parameters and type definitions
+are defined in deployment templates. Configuration object property values can be literal values such as strings,
+booleans, and objects. Values can also be directives, such as to fetch a value from an ARM resource API. A directive
+consists of one or more evaluation steps. The extensions property supplied to the deployment PUT will support language
+expressions that will be evaluated against the template defining the deployment. This allows for use of non-secure
+deployment parameters to fill in portions of the configuration (namespace, resource names). The Deployment engine will
+be responsible for validation of expressions to ensure secret values are not processed in them.
+
+main.json - The root deployment
+
+```json5
+{
+  // root deployment
+  "template": {
+    "resources": {
+      "extResources": {
+        // Nested deployment
+        "type": "Microsoft.Resources/deployments",
+        "apiVersion": "2022-09-01",
+        "name": "extResourcesDeploymentName",
+        "properties": {
+          // Language expressions can be used. Expressions are not allowed to reference secure parameters or provide data for config secrets in ways that put the secret security at risk.
+          "extensionConfigs": {
+            "k8s": {
+              "type": "object",
+              // allowed types: "object", "inherit" (more on this in a further section)
+              "value": {
+                "namespace": {
+                  "value": "[parameters('namespace')]"
+                },
+                "kubeConfig": {
+                  "value": "[listClusterAdminCredential(resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName')), '2024-02-01').kubeconfigs[0].value]",
+                  // When deployed as a stack, the value expression will be decomposed on the backend to acceptable 
+                  // formats returned by the deployment GET response. More on this in a further section.
+                }
+              }
+            }
```

Comment:
My appologies, I thought I left a comment earlier, but maybe I forgot to submit it 😑...

I'm considering whether we could simplify this by avoiding type as a discriminator.

For example:

```jsonc
// Explicitly specify the config object
{
  "k8s": {
    "namespace": {
      "value": "..."
    },
    "kubeConfig": {
      "value": "..."
    }
  }
}
```

```jsonc
// Inherit from the parent template
{
  "k8s": "k8s" // Use the extension symbol name from the parent template
}
```
The JToken type of "k8s" (`object` vs. `string`) should be enough to indicate whether the extension config object is explicitly defined or inherited.

Alternatively, instead of using symbolic name for inheritance, we could introduce an extensionConfigs function to access extension configurations directly:

```jsonc
{
  "k8s": "[extensionConfigs('k8s')]"
}
```
This approach gives users finer control over individual configuration items:

```jsonc
{
  "k8s": {
    "kubeConfig": {
      "value": "[extensionConfigs('k8s').kubeConfig]"
    },
    "namespace": {
      "value": "anotherNamespace"
    }
  }
}
```

What do you think?

Reply:
I think an `extensionsConfig` function would be useful and necessary, otherwise there's no way to combine a single kube config with multiple namespaces for example. This highlights a gap in this current design that should probably be addressed sooner than later. We don't have a way to inherit and override a subset of values yet, or conditionally inherit an extension config as a whole.

I'm not really a fan of reducing the type of the root object of an extension's config to JToken (string & object). My preference there would be to keep things as strongly typed as possible for known schemas such as the reference models.

I see object spreads codegen to something like this:
```json5
"value": "[shallowMerge(createArray(variables('objA'), createObject('prop1', 'val')))]"
```

With the current implementation doing validation at preprocessing time to check secure data usage and support API reference conversion, it feels like something similar to the object spread syntax above complicates the preprocessing logic. It seems we'd need to optimistically evaluate a statement like this to validate the overridden parts. Runtime references would prevent optimistic evaluation so analyzing the function stack becomes more complex.

What do you think of the below model?

```json5
{
  "k8s": {
    "type": "inherit",
    "alias": "[if(..., 'k8s1', 'k8s2')]",
    "value": { // for inherits, this would override portions, preprocessing would only check this
      "namespace": {
        "value": "otherNamespace"
      },
      "otherProp": {
        "value": "[extensionConfigs('k8s1').otherProp]"
      }
    }
  }
}
```

This does feel like it would complicate the Bicep codegen though to produce this alternative output for object spreads and the alias expression...

Reply:
Thinking about the model I suggested some more.. I don't think we'd support spreading reference sourced objects onto an extension config because we won't know the properties until the resource job runs. I guess if we wanted that we could possibly rerun the preprocessing like validation in resource job after references are resolvable. This syntax also doesn't support spreading an object sourced from a parameter, which depending on how complicated extension configurations could get, might be something that is desirable to do. If we wanted to support that we'd need another way to express that operation.

Reply:
@kalbert312 Would it help if you were able to leverage the folding evaluator we recently added to support what-if? Simplifying expressions that aren't fully evaluable might allow Stacks to support more complex expressions.

Reply:
@jeskew Possibly. I haven't had a chance to familiarize myself with it yet.

Reply:
Bicep discussions 11/4:
- Leave as is for now as a stopgap to a purely LE approach. The challenge is that we must know the root object properties that will be embedded in a LE without reducing the property value LEs so they can be validated.
- Add support for inheritance with overrides.
- Remove/rename confusing "type" property on the declaration.